Migrate access control example to documentation website (#6572)
nginxinc/kubernetes-ingress: site/content/configuration/access-control.md

---
title: Deploy an Access Control policy
weight: 900
toc: true
docs: DOCS-000
---

> This topic describes how to apply and update an Access Control policy with F5 NGINX Ingress Controller.

Author: I'll think a bit about this - the challenge around it is that policy exists as a general noun, whereas a Policy is a resource in context. 

There's technically no such thing as an "Access Control Policy" in terms of specificity, so I might change this to something along the lines of "Managing access control using a Policy".

> This topic describes how to apply and update an Access Control policy with F5 NGINX Ingress Controller.

Reviewer: This is a nit tbh in this PR as theres instances both ways in the examples folder. But yeah its something to think about.

It demonstrates this using an example application and a [VirtualServer custom resource]({{< ref "/configuration/virtualserver-and-virtualserverroute-resources.md" >}})

---

## Before you begin

You should have a [working NGINX Ingress Controller]({{< ref "/installation/installing-nic/installation-with-helm.md" >}}) instance.

For ease of use in shell commands, set two shell variables:

1. The public IP address for your NGINX Ingress Controller instance.

```shell
IC_IP=<ip-address>
```

> 2. The HTTP port of the same instance

Reviewer 2: ```suggestion
2. The HTTP port of the same instance.
```

```shell
IC_HTTP_PORT=<port number>
```

---

## Deploy the example application

Create the file *webapp.yaml* with the following contents:

{{< ghcode "https://raw.githubusercontent.com/nginxinc/kubernetes-ingress/refs/heads/main/examples/custom-resources/access-control/webapp.yaml" >}}

> Apply it using *kubectl*:

Reviewer 2: Should 'kubectl' be formatted as code instead?

> Apply it using *kubectl*:

Author: ```suggestion
Apply it using 'kubectl':
```

```shell
kubectl apply -f webapp.yaml
```

---

> ## Deploy an Access Control policy

Reviewer 2: ```suggestion
## Deploy a Policy to deny application access
```

Author: The focus of the document as a whole is to show how to deploy an Access Control policy: the deny and allow steps existence in sequence to demonstrate this, so I don't think it makes sense to make the policy the focus nor is it necessary to be this explicit.

Reviewer 2: You're using the same text as the title of the document in a sub-heading, which is confusing for users. If the broad goal of the doc is to deploy an access control policy, and I complete this section, am I now done with the doc? Why are there more sections? It makes more sense to me to have the sub-headings accurately reflect the expected outcome of completing the steps in the section.

Create a file named *access-control-policy-deny.yaml*. The highlighted *deny* field will be used by the example application, and should be changed to the subnet of your machine.

{{< ghcode "https://raw.githubusercontent.com/nginxinc/kubernetes-ingress/refs/heads/main/examples/custom-resources/access-control/access-control-policy-deny.yaml" "hl_lines=7-8" >}}

Apply the policy:

```shell
kubectl apply -f access-control-policy-deny.yaml
```

---

## Configure load balancing

> Create a file named *virtual-server.yaml* for the VirtualServer resource. The *policies* field references the example application.

Reviewer 2: ```suggestion
Create a file named *virtual-server.yaml* for the VirtualServer resource. The *policies* field references the access control Policy created in the previous section.
```

{{< ghcode "https://raw.githubusercontent.com/nginxinc/kubernetes-ingress/refs/heads/main/examples/custom-resources/access-control/virtual-server.yaml" "hl_lines=7-8" >}}

Apply the policy:

```shell
kubectl apply -f virtual-server.yaml
```

---

## Test the example application

Use *curl* to attempt to access the application:

```shell
curl --resolve webapp.example.com:$IC_HTTP_PORT:$IC_IP http://webapp.example.com:$IC_HTTP_PORT
```
```text
<html>
<head><title>403 Forbidden</title></head>
<body>
<center><h1>403 Forbidden</h1></center>
</body>
</html>
```

The *403* response is expected, successfully blocking your machine.

---

> ## Update the policy

Author: @vepatel Within the context of Kubernetes configuration, I have always thought of a Policy as the live configuration being updated by YAML files.

Is it more accurate to describe the YAML files themselves as separate Policies? The YAML files are ephemeral, so it feels odd to me.

> ## Update the policy

Reviewer 2: We are not asking the user to apply an update to their policy here; we're asking them to apply a new policy file, with a different name. Just from that standpoint, the heading of "update the policy" is misleading. Aside from that, the subheading should accurately reflect what the user is being instructed to do. In this case, that is deploying a policy to allow traffic. 

This also helps with SEO, so people who are searching specifically for how to deploy an allow or deny policy will easily find the section with the right instructions.

> ## Update the policy

Reviewer 3: Just to clarify, the Policy here is called `webapp-policy` so even though the updates are in two separate files, only one Policy is being updated.
